File: android/spacex/src/main/kotlin/com/openland/spacex/store/collectMissingKeys.kt
package com.openland.spacex.store

import com.openland.spacex.OutputType
import com.openland.spacex.Selector
import org.json.JSONObject

private fun collectMissingKeys(cacheKey: String, store: RecordStore, selectors: List<Selector>, arguments: JSONObject): Set<String> {
    if (!store.isInMemory(cacheKey)) {
        return setOf(cacheKey)
    }
    val record = store.read(cacheKey)

    var res = emptySet<String>()
    for (f in selectors) {
        if (f is Selector.Field) {
            val key = selectorKey(f.name, f.arguments, arguments)
            if (record.fields.containsKey(key)) {
                res += collectMissingKeys(record.fields[key]!!, f.type, store, arguments)
            }
        } else if (f is Selector.TypeCondition) {
            if (record.fields["__typename"] == RecordValue.String(f.type)) {
                res += collectMissingKeys(cacheKey, store, f.fragment.selectors, arguments)
            }
        } else if (f is Selector.Fragment) {
            res += collectMissingKeys(cacheKey, store, f.fragment.selectors, arguments)
        } else {
            error("Unsupported selector")
        }
    }
    return res
}

private fun collectMissingKeys(value: RecordValue, type: OutputType, store: RecordStore, arguments: JSONObject): Set<String> {
    if (type is OutputType.Scalar) {
        return emptySet()
    } else if (type is OutputType.NotNull) {
        return if (value == RecordValue.Null) {
            emptySet()
        } else {
            collectMissingKeys(value, type.inner, store, arguments)
        }
    } else if (type is OutputType.List) {
        if (value == RecordValue.Null) {
            return emptySet()
        } else {
            if (value is RecordValue.List) {
                var res = emptySet<String>()
                val mapped = value.items.map { collectMissingKeys(it, type.inner, store, arguments) }
                for (m in mapped) {
                    res += m
                }
                return res
            } else {
                error("Invalid record value")
            }
        }
    } else if (type is OutputType.Object) {
        if (value == RecordValue.Null) {
            return emptySet()
        } else {
            if (value is RecordValue.Reference) {
                return collectMissingKeys(value.key, store, type.selectors, arguments)
            } else {
                error("Invalid record value")
            }
        }
    }
    return emptySet()
}

fun collectMissingKeys(cacheKey: String, store: RecordStore, type: OutputType.Object, arguments: JSONObject): Set<String> {
    return collectMissingKeys(cacheKey, store, type.selectors, arguments)
}

fun collectMissingKeysRoot(rootCacheKey: String, store: RecordStore, type: OutputType.Object, arguments: JSONObject): Set<String> {
    var res = emptySet<String>()
    for (f in type.selectors) {
        if (f !is Selector.Field) {
            error("Root query cant't contain fragments")
        }
        val key = selectorKey(f.name, f.arguments, arguments)
        val refId = "$rootCacheKey.\$ref.$key"
        if (!store.isInMemory(refId)) {
            res += refId
        } else {
            val value = store.read(refId)
            val ex = value.fields["data"]
            if (ex != null) {
                res += collectMissingKeys(ex, f.type, store, arguments)
            }
        }
    }
    return res
}

File: packages/openland-graphql/spacex/store/collectMissingKeys.ts
import { RecordStore, RecordValue } from './RecordStore';
import { OutputTypeObject, OutputType, Selector } from './../types';
import { selectorKey } from './selectorKey';

function collectMissingKeysFromSelectors(cacheKey: string, store: RecordStore, selectors: Selector[], variables: any, res: Set<string>) {
    if (!store.isInMemory(cacheKey)) {
        res.add(cacheKey);
        return;
    }
    let record = store.read(cacheKey);

    for (let f of selectors) {
        if (f.type === 'field') {
            let key = selectorKey(f.name, f.arguments, variables);
            if (record.fields[key] !== undefined) {
                collectMissingKeysFromValue(record.fields[key], f.fieldType, store, variables, res);
            }
        } else if (f.type === 'type-condition') {
            if (record.fields.__typename && record.fields.__typename.type === 'string' && record.fields.__typename.value === f.type) {
                collectMissingKeysFromSelectors(cacheKey, store, f.fragmentType.selectors, variables, res);
            }
        } else if (f.type === 'fragment') {
            collectMissingKeysFromSelectors(cacheKey, store, f.fragmentType.selectors, variables, res);
        } else {
            throw Error('Unreachable code');
        }
    }
}

function collectMissingKeysFromValue(value: RecordValue, type: OutputType, store: RecordStore, variables: any, res: Set<string>) {
    if (type.type === 'scalar') {
        return;
    } else if (type.type === 'notNull') {
        if (value.type !== 'null') {
            collectMissingKeysFromValue(value, type.inner, store, variables, res);
        }
    } else if (type.type === 'list') {
        if (value.type === 'list') {
            for (let v of value.values) {
                collectMissingKeysFromValue(v, type.inner, store, variables, res);
            }
        } else if (value.type !== 'null') {
            throw Error('Invalid record value');
        }
    } else if (type.type === 'object') {
        if (value.type === 'reference') {
            collectMissingKeysFromSelectors(value.key, store, type.selectors, variables, res);
        } else if (value.type !== 'null') {
            throw Error('Invalid record value');
        }
    }
}

export function collectMissingKeysRoot(rootCacheKey: string, store: RecordStore, type: OutputTypeObject, variables: any): Set<string> {
    let res = new Set<string>();
    for (let f of type.selectors) {
        if (f.type !== 'field') {
            throw Error('Root query cant\'t contain fragments');
        }
        let key = selectorKey(f.name, f.arguments, variables);
        let refId = rootCacheKey + '.$ref.' + key;
        if (!store.isInMemory(refId)) {
            res.add(refId);
        } else {
            let value = store.read(refId);
            let ex = value.fields.data;
            if (ex) {
                collectMissingKeysFromValue(ex, f.fieldType, store, variables, res);
            }
        }
    }
    return res;
}